### PR TITLE
Use sets for dependencies inside the engine

### PR DESCRIPTION
**Pull request by Frassle**
# Description

We're doing more and more manipulation of these URN slices. We should just make these sets to make it simpler to manipulate them without worrying about order or duplicates.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/resource/deploy/source_eval.go [hunk 1]
```diff
@@ -2147,30 +2155,29 @@ func downgradeOutputValues(v resource.PropertyMap) resource.PropertyMap {
 	return result
 }
 
-func addOutputDependencies(deps []resource.URN, v resource.PropertyValue) []resource.URN {
+func addOutputDependencies(deps mapset.Set[resource.URN], v resource.PropertyValue) {
```
Comment (PollRobots) [inline pkg/resource/deploy/source_eval.go:2158]: At every external callsite the set is constructed from a slice, then addOutputDependencies is called on it, and then the set is converted back to a slice.

Would it be cleaner to have a wrapper function for `addOutputDependencies` that took a slice of URNS, and a propertyValue and returned a slice of URNs, then have that contain the pattern. 

something like

``` go
func extendOutputDependencies(existing []resource.URN, v resource.PropertyValue) []resource.URN {
    deps := mapset.NewSet(existing...)
    addOutputDependencies(deps, v)
    return deps.ToSlice()
}
```

**Review by PollRobots — APPROVED** [review]
LGTM

I'm assuming that we don't care about the ordering here...

One minor comment

**Comment by Frassle** [comment]
> I'm assuming that we don't care about the ordering here...

We shouldn't. I'm going to add a test to confirm that we're not getting duplicates though.

File: pkg/resource/deploy/source_eval.go (hunk 1 above)
Comment (Frassle) [reply pkg/resource/deploy/source_eval.go:2158]: Maybe, I'll recheck how this interacts with the other changes I'm doing that use the sets even more.

**Comment by Frassle** [comment]
I'll rebase this ontop of https://github.com/pulumi/pulumi/pull/15675 with a test that we don't duplicate deps (currently we can)

File: pkg/resource/deploy/source_eval.go (hunk 1 above)
Comment (Frassle) [reply pkg/resource/deploy/source_eval.go:2158]: Yeh this looks reasonable, I'll add this
